src/borrow_tracker/tree_borrows/tree.rs

```diff
@@ -0,0 +1,548 @@
+use smallvec::SmallVec;
+
+use rustc_const_eval::interpret::InterpResult;
+use rustc_data_structures::fx::FxHashSet;
+use rustc_target::abi::Size;
+
+use crate::borrow_tracker::tree_borrows::{
+    diagnostics::{NodeDebugInfo, TbError, TransitionError, TransitionResult},
+    unimap::{UniEntry, UniIndex, UniKeyMap, UniValMap},
+    Permission,
+};
+use crate::borrow_tracker::{GlobalState, ProtectorKind};
+use crate::*;
+
+/// Records whether the location was ever directly accessed,
+/// or if it was only initialized because of a foreign access.
+// In the latter case any protector would not trigger.
+// FIXME: rethink this
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub(super) struct AccessChk<T> {
+    /// The wrapped value
+    pub data: T,
+    /// Whether the value was ever accessed
+    pub accessed: bool,
+    /// Latest transition that was applied to this node:
+    /// Some(b) if foreign, with b=true for foreign write and b=false for foreign read;
```

Comment:
This sounds like it should use an `enum`, not `bool`. We do have an `AccessKind` enum somewhere.